### PR TITLE
List volumes by service offering id

### LINKED ISSUE #8646: List volume api should have the parameter service offering id
ISSUE TYPE

Improvement request

COMPONENT NAME

Component: Api

CLOUDSTACK VERSION

Cloudstack version 4.18 4.19

SUMMARY

List volume api should have the option of service offering id 

Steps to reproduce the issue

1. Create a service offering 

2. Launch a vm with service offering 

3. Currently if I want to list the root disk volumes  of the vm created from the service offering 

There is no way list them

https://cloudstack.apache.org/api/apidocs-4.19/apis/listVolumes.html

It would be handy feature for the admins and user to have a servicofferingid parameter as they could list all the volumes associated with the service offering id 

Example 

list volume <serviceoffering id>

This is could be similar to listVirtualMachine with serviceofferingid parameter

https://cloudstack.apache.org/api/apidocs-4.19/apis/listVirtualMachines.html

Root disk details, no way to identify the service offering

Screenshot 2024-02-13 at 3 42 04 PM

Data disk details 

data disk details

### PR DESCRIPTION
**Pull request by vishesh92**
### Description

Fixes #8646 
This PR allows listing of volumes by service offering id. If both service offering id and disk offering id are passed, service offering id is ignored.
UI is also updated to allow filtering of volumes by service offering id and disk offering id.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### Screenshots (if appropriate):
image

### CODE REVIEW

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan test keepEnv

**Review by rohityadavcloud — APPROVED** [review]

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 1]
```diff
@@ -2416,6 +2417,13 @@ private Pair<List<Long>, Integer> searchForVolumeIdsAndCount(ListVolumesCmd cmd)
 
         List<Long> ids = getIdsListFromCmd(cmd.getId(), cmd.getIds());
 
+        if (diskOffId == null && serviceOfferingId != null) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:2420]: _diskOffId_ => _diskOfferingId_

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by kiranchavala — APPROVED** [review]
LGTM, Able to list the volumes based on service offering id and disk offering id 

Screenshot 2024-06-12 at 3 27 44 PM

Screenshot 2024-06-12 at 3 28 02 PM


Also the list volumes api call has the parameter serviceofferingid

```

(local2) 🐱 > list volumes serviceofferingid=69e69db8-021d-46bb-acfc-bdce0761f9e5
{
  "count": 1,
  "volume": [
    {
      "account": "user",
      "created": "2024-06-12T15:20:24+0530",
      "destroyed": false,
      "deviceid": 0,
      "diskioread": 0,
      "diskiowrite": 0,
      "diskkbsread": 0,
      "diskkbswrite": 0,
      "domain": "ROOT",
      "domainid": "ae0b887b-2895-11ef-8e34-74563c46f986",
      "hasannotations": false,
      "id": "4acf971f-6adf-4ed9-9c5b-51af4b60cedb",
      "isextractable": false,
      "name": "ROOT-5",
      "provisioningtype": "thin",
      "quiescevm": false,
      "serviceofferingdisplaytext": "Medium Instance",
      "serviceofferingid": "2",
      "serviceofferingname": "Medium Instance",
      "size": 2147483648,
      "state": "Ready",
      "supportsstoragesnapshot": false,
      "tags": [],
      "templatedisplaytext": "CentOS 5.6 (64-bit) no GUI (Simulator)",
      "templateid": "c182a3cd-2895-11ef-8e34-74563c46f986",
      "templatename": "CentOS 5.6 (64-bit) no GUI (Simulator)",
      "type": "ROOT",
      "virtualmachineid": "2dcd0954-80aa-4a78-924f-714bea1a0e00",
      "vmdisplayname": "VM-2dcd0954-80aa-4a78-924f-714bea1a0e00",
      "vmname": "VM-2dcd0954-80aa-4a78-924f-714bea1a0e00",
      "vmstate": "Running",
      "vmtype": "User",
      "zoneid": "770dc1d7-3381-4b22-95b6-47f1bf1c5f0f",
      "zonename": "Sandbox-simulator"
    }
  ]
}

```

**Comment by sureshanaparti** [comment]
Merging on the tests and reviews done.

**Pull request merged**